cmd/kubectl-gadget/utils/trace.go

```diff
@@ -104,169 +134,304 @@ func deleteTraces(contextLogger *log.Entry, traceRestClient *restclient.RESTClie
-				},
-				RunMode:    "Manual",
-				OutputMode: outputMode,
-			},
+		_, present := trace.ObjectMeta.Annotations[GADGET_OPERATION]
+		if !present {
+			return fmt.Errorf("Trace with ID %q does not contain key %q", traceID, GADGET_OPERATION)
 		}
 
-		err = traceRestClient.
-			Post().
-			Namespace(trace.ObjectMeta.Namespace).
-			Resource("traces").
-			Body(trace).
-			Do(context.TODO()).
-			Error()
+		trace.ObjectMeta.Annotations[GADGET_OPERATION] = operation
+
+		err = updateTraceOperation(&trace)
 		if err != nil {
-			deleteTraces(nil, traceRestClient, traceID)
-			contextLogger.Fatalf("Error creating trace on node %s: %q", node.Name, err)
+			return err
 		}
 	}
 
-	if params.Node != "" && !nodeFound {
-		contextLogger.Fatalf("Invalid filter: Node %q does not exist", params.Node)
-	}
+	return nil
+}
 
-	var listTracesOptions = metav1.ListOptions{
-		LabelSelector: fmt.Sprintf("trace-template-hash=%s", traceID),
-		FieldSelector: fields.Everything().String(),
+func reconstructParamsFromTrace(trace *gadgetv1alpha1.Trace) *CommonFlags {
+	return &CommonFlags{
+		OutputMode: trace.Spec.OutputMode,
+		Node:       trace.Spec.Node,
 	}
+}
 
-	// Wait until results are ready and fetch all results
+// PrintTraceOutput is used to print trace output.
+// Note that, in case of gadget based-on the monitor/report operations, this
+// function must be called after trace operation was set to report.
+func PrintTraceOutput(traceID string, verbose bool,
+	customResultsDisplay func(nodes *corev1.NodeList, results *gadgetv1alpha1.TraceList),
+	transformLine func(string) string) error {
 	// TODO: use a watcher to avoid looping client-side
-	//
-	// watch, err := traceRestClient.
-	//	Get().
-	//	Namespace("gadget").
-	//	Resource("traces").
-	//	VersionedParams(&listTracesOptions, scheme.ParameterCodec).
-	//	Watch(context.TODO())
-	// if err != nil {
-	//	contextLogger.Fatalf("Error waiting for traces: %q", err)
-	// }
-	// for event := range watch.ResultChan() {
-	//	fmt.Printf("Event: %v\n", event)
-	//	if data, ok := event.Object.(*metav1.Status); ok {
-	//		contextLogger.Infof("watcher status: %s", data.Message)
-	//	} else if t, ok := event.Object.(*gadgetv1alpha1.Trace); ok {
-	//		fmt.Printf("Got: %v\n", t)
-	//	} else {
-	//		contextLogger.Fatalf("Error waiting for traces: got unexpected %v", event)
-	//	}
-	// }
-
-	var results gadgetv1alpha1.TraceList
+	var err error
+	var trace *gadgetv1alpha1.Trace
+	var traces gadgetv1alpha1.TraceList
+
 	start := time.Now()
 RetryLoop:
 	for {
-		results = gadgetv1alpha1.TraceList{}
-		err = traceRestClient.
-			Get().
-			Namespace("gadget").
-			Resource("traces").
-			VersionedParams(&listTracesOptions, scheme.ParameterCodec).
-			Do(context.TODO()).
-			Into(&results)
+		traces, err = getTraceListFromID(traceID)
 		if err != nil {
-			deleteTraces(contextLogger, traceRestClient, traceID)
-			contextLogger.Fatalf("Error getting traces: %q", err)
+			return fmt.Errorf("Error getting traces: %w", err)
```

Comment:
This error doesn't add anything on top of the error message already returned by getTraceListFromID.
```suggestion
			return err
```